mm2src/coins/qrc20/history.rs

```diff
@@ -548,169 +530,134 @@ impl Qrc20Coin {
     }
 }
 
-pub struct HistoryBuilder {
+pub struct TransferHistoryBuilder {
     coin: Qrc20Coin,
     from_block: u64,
-    topic: String,
     address: H160,
     token_address: H160,
-    order: Option<HistoryOrder>,
-}
-
-pub struct HistoryCont<T> {
-    history: Vec<T>,
 }
 
-/// Future for the [`HistoryBuilder::build_utxo_lazy()`].
-/// Loads `UtxoTx` from `tx_hash`.
-pub struct UtxoFromHashFuture {
-    future: Option<Box<dyn Future<Output = Result<UtxoTx, String>> + Unpin + 'static>>,
-}
-
-pub enum HistoryOrder {
-    NewestToOldest,
-    OldestToNewest,
-}
-
-impl HistoryBuilder {
-    pub fn new(coin: Qrc20Coin) -> HistoryBuilder {
-        let address = qrc20_addr_from_utxo_addr(coin.utxo.my_address.clone());
+impl TransferHistoryBuilder {
+    pub fn new(coin: Qrc20Coin) -> TransferHistoryBuilder {
+        let address = qtum::contract_addr_from_utxo_addr(coin.utxo.my_address.clone());
         let token_address = coin.contract_address;
-        HistoryBuilder {
+        TransferHistoryBuilder {
             coin,
             from_block: 0,
-            topic: QRC20_TRANSFER_TOPIC.to_string(),
             address,
             token_address,
-            order: None,
         }
     }
 
     #[allow(clippy::wrong_self_convention)]
-    pub fn from_block(mut self, from_block: u64) -> HistoryBuilder {
+    pub fn from_block(mut self, from_block: u64) -> TransferHistoryBuilder {
         self.from_block = from_block;
         self
     }
 
-    #[allow(dead_code)]
-    pub fn topic(mut self, topic: &str) -> HistoryBuilder {
-        self.topic = topic.to_string();
-        self
-    }
-
-    pub fn address(mut self, address: H160) -> HistoryBuilder {
+    pub fn address(mut self, address: H160) -> TransferHistoryBuilder {
         self.address = address;
         self
     }
 
     #[allow(dead_code)]
-    pub fn token_address(mut self, token_address: H160) -> HistoryBuilder {
+    pub fn token_address(mut self, token_address: H160) -> TransferHistoryBuilder {
         self.token_address = token_address;
         self
     }
 
-    pub fn order(mut self, order: HistoryOrder) -> HistoryBuilder {
-        self.order = Some(order);
-        self
+    pub async fn build(self) -> Result<Vec<TxReceipt>, JsonRpcError> {
+        match self.coin.utxo.rpc_client {
+            UtxoRpcClientEnum::Native(ref native) => self.build_receipts_with_native(native).await,
```

Comment:
I think it's worth extracting this into the trait.

Reply:
Done